src/java/org/apache/cassandra/utils/SystemInfo.java

```diff
@@ -0,0 +1,214 @@
+            Path p =  FileSystems.getDefault().getPath("/proc", Long.toString(getPid()), "limits");
+            try
+            {
+                List<String> lines = Files.readAllLines(p);
+                for (String s : lines)
+                {
+                    if (s.startsWith("Max processes"))
+                    {
+                        String[] parts = s.split("\s+");
+                        String limit = parts[2];
+                        if ("unlimited".equals(limit))
+                        {
+                            return INFINITY;
+                        }
+                        return Long.parseLong(limit);
+                    }
+                }
+                logger.error("'Max processes' not found in " + p);
+            }
+            catch (IOException e)
+            {
+                logger.error("Unable to read "+p, e);
+            }
+        }
+        /* return the default value for non-Linux systems.
+         * can not return 0 as we know there is at least 1 process (this one) and
+         * -1 historically represents infinity.
+        */
+        return DEFAULT_MAX_PROCESSES;
+    }
+
+    /**
+     *
+     * @return The maximum number of open files allowd to the current process/user.
+     */
+    long getMaxOpenFiles()  {
+        // ulimit -H -n
+        return si.getOperatingSystem().getCurrentProcess().getHardOpenFileLimit();
+    }
+
+    /**
+     * Gets the Virtual Memory Size (VSZ). Includes all memory that the process can access, including memory that is swapped out and memory that is from shared libraries.
+     * @return The amount of virtual memory allowed to be allocatedby the current process/user.
+     */
+    long getVirtualMemoryMax() {
+        return si.getOperatingSystem().getCurrentProcess().getVirtualSize();
+    }
+
+    /**
+     *
+     * @return The amount of swap space allocated on the system.
+     */
+    long getSwapSize() {
+        return si.getHardware().getMemory().getVirtualMemory().getSwapTotal();
+    }
+
+    /**
+     * @return the PID of the current system.
+     */
+    public long getPid() {
+        return si.getOperatingSystem().getProcessId();
+    }
+
+    /**
+     * Checks if a value if valide (i.e. value >= min or value == INFINITY..
+     * @param value the value to check.
+     * @param min the minimum value.
+     * @return true if value is valid.
+     */
+    private boolean valid(long value, long min)
+    {
+        return value >= min || value == INFINITY;
+    }
+
+    /**
+     * Tests if the system is running in degraded mode.
+     * If the system is running in degraded mode this method will log information.
+     * @return @{code true} if the system is in degraded mode, @{code false} otherwise.
+     */
+    public boolean warnIfRunningInDegradedMode()
+    {
+        boolean degraded = false;
+        StringBuilder sb = new StringBuilder();
+
+        if (oshi.SystemInfo.getCurrentPlatform() == PlatformEnum.LINUX)
```

Comment:
I find this logic complicated. Ifs, elses ... degraded set to true etc ... We can do better.